File: src/main/java/cube8540/oauth/authentication/credentials/oauth/client/application/DefaultOAuth2ClientDetailsService.java
package cube8540.oauth.authentication.credentials.oauth.client.application;

import cube8540.oauth.authentication.credentials.oauth.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.OAuth2ClientDetailsService;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientRepository;
import cube8540.oauth.authentication.credentials.oauth.error.ClientNotFoundException;
import lombok.AccessLevel;
import lombok.Getter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class DefaultOAuth2ClientDetailsService implements OAuth2ClientDetailsService {

    @Getter(AccessLevel.PROTECTED)
    private final OAuth2ClientRepository repository;

    @Autowired
    public DefaultOAuth2ClientDetailsService(OAuth2ClientRepository repository) {
        this.repository = repository;
    }

    @Override
    public OAuth2ClientDetails loadClientDetailsByClientId(String clientId) {
        return repository.findByClientId(new OAuth2ClientId(clientId))
                .map(DefaultOAuth2ClientDetails::of)
                .orElseThrow(() -> ClientNotFoundException.instance(clientId + " is not found"));
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/credentials/oauth/client/application/DefaultOAuth2ClientDetailsService.kt
package cube8540.oauth.authentication.credentials.oauth.client.application

import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientRepository
import cube8540.oauth.authentication.credentials.oauth.error.OAuth2ClientRegistrationException
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetailsService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class DefaultOAuth2ClientDetailsService(private val repository: OAuth2ClientRepository): OAuth2ClientDetailsService {

    @Transactional(readOnly = true)
    override fun loadClientDetailsByClientId(clientId: String): OAuth2ClientDetails =
        repository.findByClientId(OAuth2ClientId(clientId))
            .map(DefaultOAuth2ClientDetails::of)
            .orElseThrow { OAuth2ClientRegistrationException("$clientId is not found") }
}